core/headerchain.go

```diff
@@ -591,353 +494,189 @@ func (hc *HeaderChain) CurrentHeader() *types.Header {
-	if err != nil {
-		return common.Hash{}, err
+// GetBody retrieves a block body (transactions and uncles) from the database by
+// hash, caching it if found.
+func (hc *HeaderChain) GetBody(hash common.Hash) *types.Body {
+	// Short circuit if the body's already in the cache, retrieve otherwise
+	if cached, ok := hc.bc.bodyCache.Get(hash); ok {
+		body := cached.(*types.Body)
+		return body
 	}
-
-	if RTZ.Hash() != RTR.Hash() {
-		return common.Hash{}, errors.New("there exists a region twist")
+	number := hc.GetBlockNumber(hash)
+	if number == nil {
+		return nil
 	}
-
-	// Prime twist check
-	// PTZ -- Prime coincident along zone path
-	// PTR -- Prime coincident along region path
-	// PTP -- Prime coincident along prime path
-	PTZ, err := hc.Engine().PreviousCoincidentOnPath(hc, header, slice, params.PRIME, params.ZONE, true)
-	if err != nil {
-		return common.Hash{}, err
+	body := rawdb.ReadBody(hc.headerDb, hash, *number)
+	if body == nil {
+		return nil
 	}
+	// Cache the found body for next time and return
+	hc.bc.bodyCache.Add(hash, body)
+	return body
+}
 
-	PTR, err := hc.Engine().PreviousCoincidentOnPath(hc, header, slice, params.PRIME, params.REGION, true)
-	if err != nil {
-		return common.Hash{}, err
+// GetBodyRLP retrieves a block body in RLP encoding from the database by hash,
+// caching it if found.
+func (hc *HeaderChain) GetBodyRLP(hash common.Hash) rlp.RawValue {
+	// Short circuit if the body's already in the cache, retrieve otherwise
+	if cached, ok := hc.bc.bodyRLPCache.Get(hash); ok {
+		return cached.(rlp.RawValue)
 	}
-
-	PTP, err := hc.Engine().PreviousCoincidentOnPath(hc, header, slice, params.PRIME, params.PRIME, true)
-	if err != nil {
-		return common.Hash{}, err
+	number := hc.GetBlockNumber(hash)
+	if number == nil {
+		return nil
 	}
-
-	if PTZ.Hash() != PTR.Hash() || PTR.Hash() != PTP.Hash() || PTP.Hash() != PTZ.Hash() {
-		return common.Hash{}, errors.New("there exists a prime twist")
+	body := rawdb.ReadBodyRLP(hc.headerDb, hash, *number)
+	if len(body) == 0 {
+		return nil
 	}
-
-	return PTP.Hash(), nil
+	// Cache the found body for next time and return
+	hc.bc.bodyRLPCache.Add(hash, body)
+	return body
 }
 
-// calcHLCRNetDifficulty calculates the net difficulty from previous prime.
-// The netDifficulties parameter inputs the nets of instantaneous difficulties from the terminus block.
-// By correctly summing the net difficulties we have obtained the proper array to be compared in HLCR.
-func (hc *HeaderChain) CalcHLCRNetDifficulty(terminalHash common.Hash, header *types.Header) ([]*big.Int, error) {
-
-	if (terminalHash == common.Hash{}) {
-		return nil, errors.New("one or many of the  terminal hashes were nil")
+// GetBlocksFromHash returns the block corresponding to hash and up to n-1 ancestors.
+// [deprecated by eth/62]
+func (hc *HeaderChain) GetBlocksFromHash(hash common.Hash, n int) (blocks []*types.Block) {
```

Comment:
Similar to my comment about `GetBlockHashesFromHash()` I think this should be renamed as `GetAncestorBlocks()` or something more meaningful.